linux/devdoc/threadpool_linux_requirements.md

```diff
@@ -0,0 +1,350 @@
+
+`threadpool_schedule_work` shall increment the `insert_pos`.
+
+If task state is `TASK_NOT_USED`, `threadpool_schedule_work` shall set the current task state to `TASK_INITIALIZING`.
+
+Otherwise, `threadpool_schedule_work` shall release the shared SRW lock by calling `srw_lock_release_shared` and increase `task_array` capacity by calling `reallocate_threadpool_array`.
+
+  - `reallocate_threadpool_array` shall acquire the SRW lock in exclusive mode by calling `srw_lock_acquire_exclusive`.
+
+  - `reallocate_threadpool_array` shall get the current size of task array by calling `interlocked_add`.
+
+  - If there is any overflow computing the new size, `reallocate_threadpool_array` shall fail and return a non-zero value .
+
+  - Otherwise, `reallocate_threadpool_array` shall double the current task array size.
+
+  - `reallocate_threadpool_array` shall realloc the memory used for the array items and on success return a non-`NULL` handle to it.
+
+  - If any error occurs, `reallocate_threadpool_array` shall fail and return a non-zero value.
+
+  - `reallocate_threadpool_array` shall initialize every task item in the new task array with `task_func` and `task_param` set to `NULL` and `task_state` set to `TASK_NOT_USED`.
+
+  - `reallocate_threadpool_array` shall remove any gap in the task array.
+
+  - `reallocate_threadpool_array` shall reset the `consume_idx` and `insert_idx` to 0 after resize the task array.
+
+  - `reallocate_threadpool_array` shall release the SRW lock by calling `srw_lock_release_exclusive`.
+
+  - `reallocate_threadpool_array` shall return zero on success.
+
+If reallocating the task array fails, `threadpool_schedule_work` shall fail and return a non-zero value.
+
+`threadpool_schedule_work` shall obtain task information in next available task array index and return zero on success.
+
+`threadpool_schedule_work` shall set the `task_state` to `TASK_WAITING` and then release the shared SRW lock.
+
+`threadpool_schedule_work` shall unblock the `threadpool` semaphore by calling `sem_post`.
+
+`threadpool_schedule_work` shall decrement the count of pending call that are in progress to be executed.
+
+`threadpool_schedule_work` shall call `sm_exec_end`.
+
+### threadpool_timer_start
+
+```C
+MOCKABLE_FUNCTION(, int, threadpool_timer_start, THREADPOOL_HANDLE, threadpool, uint32_t, start_delay_ms, uint32_t, timer_period_ms, THREADPOOL_WORK_FUNCTION, work_function, void*, work_function_ctx, TIMER_INSTANCE_HANDLE*, timer_handle);
+```
+
+`threadpool_timer_start` starts a threadpool timer which runs after `start_delay_ms` milliseconds and then runs again every `timer_period_ms` milliseconds until `threadpool_timer_cancel` or `threadpool_timer_destroy` is called. The `timer_handle` must be stopped before closing/destroying the threadpool.
+
+If `threadpool` is `NULL`, `threadpool_timer_start` shall fail and return a non-zero value.  
```

Comment:
>   [](http://example.com/codeflow?start=91&length=2)

Nitpick: extra spaces at the end. There is a vscode addon AFAIK that helps with highlighting these. Matt knows :-)

Reply:
done